docs/error_handling.md

```diff
@@ -0,0 +1,200 @@
+
+```nim
+proc foo(x: int, y: string): float {.errors: (ValueError, KeyError).} =
+  body
+```
+
+is re-written to the equivalent of:
+
+```nim
+type
+  Raising[ErrorsList, ResultType] = distinct ResultType
+
+proc foo_original(x: int, y: string): float {.
+  raises: [Defect, ValueError, KeyError]
+.} =
+  body
+
+template foo(x: int, y: string): untyped =
+  Raising[(ValueError, KeyError), float](foo_original(x, y))
+```
+
+Please note that the original proc now features a `raises` annotation
+that will guarantee that no other exceptions might be raised from it.
+The `Defect` type was implicitly added to the list as a convenience.
+
+The returned distinct type will be useless at the call-site unless
+it is stripped-away through `raising`, `either` or `check` which are
+the error-handling mechanisms provided by this library and discussed
+further in this document.
+
+If you accidentally forget to use one of the error-handling mechanisms,
+you'll get a compilation error along these lines:
+
+```
+required type for x: float
+  but expression 'Raising[(ValueError, KeyError), float](foo_original(x, y))' is of type: Raising[tuple of (ValueError, KeyError), system.float]
+```
+
+Please note that if you have assigned the `Raising` result to a
+variable, the compilation error might happen on a line where you
+attempt to use that variable. To fix the error, please introduce
+error handling as early as possible at the right call-site such
+that no `Raising` variable is created at all.
+
+`noerrors` is another pragma provided for convenience which is
+equivalent to an empty `errors` pragma. The forced error handling
+through the `Raising` type won't be applied.
+
+Both pragmas can be combined with the `nodefects` pragma that
+indicates that the specific proc should be proven to be Defect-free.
+
+The transformation uses a template by default to promote efficiency,
+but if you need to take the address the Raising proc, please add the
+`addressable` pragma that will force the wrapper to be a regular proc.
+
+Finally, `failing` is another pragma provided for convenience which
+is equivalent to `{.errors: (CatchableError).}`.
+
+## The `raising` annotation
+
+The `raising` annotation is the simplest form of error tracking
+similar to the `try` annotation made popular by the [Midori error model](http://joeduffyblog.com/2016/02/07/the-error-model/),
+which is also [proposed for addition in C++](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0709r0.pdf) and [already available in Zig](https://ziglang.org/documentation/master/#try).
+
+It merely marks the locations in the code where errors might be raised
+and strips away the `Raising` type to disarm the compiler checks:
+
+```nim
+proc attachValidator(node: BeaconNode, keyFile: string) {.failing.} =
+  node.addValidator(raising ValidatorPrivKey.init(raising readFile(keyFile))
+```
+
+When applied to a `Result` or an `Option`, `raising` will use the
+`tryGet` API to attempt to obtain the computed value.
```

Comment:
for unification with `Result`, this is practically the same as `?` - should `?` be smarter and convert back-and-forth between the error handling models? from a performance POV, that's a bit dumb (the visual cost is much smaller than the incurred penalty) but it's... pragmatic.

Reply:
I considered this as well, but it may be indeed too magical. I guess we'll know soon with more practice.